### PR TITLE
Feature/gemify

### LINKED ISSUE #56: App manifest file to configure environment-specific parameters instead of ENV per branch

### LINKED ISSUE #89: Gemify deploy-script
`zen-deployer`??

- [ ] Default to config file
- [ ] Allow binary parameter overrides

Keep this private/internal until we refactor steps into modular tasks.

### LINKED ISSUE #91: Refactor script to modular tasks
Do this after #89 .

Something like Railties would be interesting for this.

### LINKED ISSUE #104: Allow promoting to multiple piplelines
We have a need to have more than one production-like environment, but the script currently assumes one.

### PR DESCRIPTION
**Pull request by gabrieljoelc**
Closes #104 
Closes #91 
Closes #89 
Closes #56

### CODE REVIEW

File: lib/zen/deploy/steps/check_local_tasks_supported.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class CheckLocalTasksSupported < Step
+        def perform(_deployer, _config)
+          File.read('Gemfile').include?('after_party')
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/check_local_tasks_supported.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/check_remote_tasks_supported.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class CheckRemoteTasksSupported < Step
+        def perform(_deployer, config)
+          output = Command.run("heroku run bundle show after_party -a#{config.app_name}", show_err: false, show_out: config.debug).out
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/check_remote_tasks_supported.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/disable_preboot.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class DisablePreboot < Step
+        def perform(deployer, config)
+          return unless deployer.captured_steps[:check_for_new_migrations]
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/disable_preboot.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/get_local_schema_versions.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class GetLocalSchemaVersions < Step
+        def perform(_deployer, config)
+          Command.run('rails r "puts ActiveRecord::Migrator.get_all_versions"', show_err: false, show_out: config.debug).out
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/get_local_schema_versions.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/get_remote_schema_versions.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class GetRemoteSchemaVersions < Step
+        def perform(_deployer, config)
+          Command.run("heroku run rails r \\\"puts ActiveRecord::Migrator.get_all_versions\\\" -a#{config.app_name} --exit-code").out
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/get_remote_schema_versions.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/get_remote_task_versions.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class GetRemoteTaskVersions < Step
+        def perform(deployer, config)
+          return unless deployer.captured_steps[:check_remote_tasks_supported]
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/get_remote_task_versions.rb]: need to check `unsafe_deploy`

File: lib/zen/deploy/steps/disable_preboot.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+          command = Command.run "heroku features:disable -a#{config.app_name} preboot"
+          deployer.fatal_error 'There was a problem disabling preboot' if command.status != 0
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/disable_preboot.rb]: Ignore "already disabled" error

File: lib/zen/deploy.rb [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      def fatal_error(msg)
+        run_ensure_steps if defined?(run_ensure_steps)
+        notify_slack "_ERROR_: #{msg}", 'danger'
+        Scriptster.log :err, msg
+        exit 1
+      end
+
+      private
+
+      def run_deployment(deployment)
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: Remove this

File: lib/zen/deploy.rb [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        run_steps(:steps, deployment)
+      end
+
+      def run_ensure(deployment)
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: Remove this

File: lib/zen/deploy.rb [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        run_steps(:ensure, deployment)
+      end
+
+      def run_steps(group, deployment)
+        deployment_config = OpenStruct.new(config.to_h.merge(deployment.to_h))
+        @slack_notifier = SlackNotifier.new(deployment_config)
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: Make this a private getter/accessor

File: lib/zen/deploy.rb [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            fatal_error "#{step} is not defined."
+          end
+        end
+      end
+
+      def run_step(step, deployment_config)
+        step_class = "Zen::Deploy::Steps::#{step}".constantize
+        log_debug "Starting step #{step}"
+        step_i = step_class.new
+        output = step_i.perform(self, deployment_config)
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: Replace invoking `fatal_error` anywhere outside of this class with `raise StepError` and `raise ConfigurationError`, respectively.

File: lib/zen/deploy/steps/capture_backup.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+require 'zen/deploy/steps/step'
+module Zen
+  module Deploy
+    module Steps
+      class CaptureBackup < Step
+        def perform(deployer, config)
+          return unless deployer.captured_steps[:check_for_new_migrations] || deployer.captured_steps[:check_for_new_tasks]
```
Comment (gabrieljoelc) [inline lib/zen/deploy/steps/capture_backup.rb]: Check `unsafe_deploy`

File: lib/zen/deploy.rb [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        end
+      end
+
+      def notify_slack(msg, color = '#ccc')
+        @slack_notifier.notify(msg, color) if @slack_notifier.present?
+      end
+
+      def fatal_error(msg)
+        run_ensure_steps if defined?(run_ensure_steps)
+        notify_slack "_ERROR_: #{msg}", 'danger'
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: it's not appending to the message

File: lib/zen/deploy.rb [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+          @captured_steps = {}.with_indifferent_access
+          verify_steps(deployment.steps)
+          if deployment.ensure.present?
+            verify_steps(deployment.ensure)
+            define_singleton_method 'run_ensure_steps' do
+              run_ensure(deployment)
+            end
+          end
+          log_debug "Starting deployment of #{deployment[:app_name]} (#{index + 1} of #{config.deployments.size})"
+          run_deployment(deployment)
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: pls to prepend current deployment to output
Scriptster::configure do |conf|
  conf.name = "my-deployment"
end

File: lib/zen/deploy.rb [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+          end
+        end
+      end
+
+      def run_step(step, deployment_config)
+        step_class = "Zen::Deploy::Steps::#{step}".constantize
+        log_debug "Starting step #{step}"
+        step_i = step_class.new
+        output = step_i.perform(self, deployment_config)
+        @captured_steps[step.underscore] = output if step_i.save_output?
```
Comment (gabrieljoelc) [inline lib/zen/deploy.rb]: also add:
```ruby
step_class.steps_that_are_stopped.each do |stopped_stop|
  @captured_steps["stop_#{stopped_stop.name.underscore}"] = output
end
```
Maybe make a `StepStopperStep < Step` base class for all steps that can stop future steps?

**Comment by gabrieljoelc** [comment]
I left all the "check `unsafe_deploy`" comments but we are actually going to remote `unsafe_deploy` in favor of "stop step" steps.

**Comment by gabrieljoelc** [comment]
- [ ] Remove unsafe deployment

**Comment by gabrieljoelc** [comment]
- [ ] Rename `ensure` to `rescue`
- [ ] Possibly add `pre_steps`

**Comment by gabrieljoelc** [comment]
From @dukedave:
> @crazydog115 is there a roadmap for how these will be set?
> It just feels a bit weird to be checking in something like:
> 
> ```
> slack_api_env_var: 'SLACK_API_TOKEN'
> ```

Why does it feel weird? Sidekiq does [it](https://github.com/mperham/sidekiq/wiki/Using-Redis#using-an-env-variable).

**Comment by crazydog115** [comment]
Yeah I don't think it's weird at all. It's the only way to set the slack api token without checking it in to the repo, or just having a magic environment variable name.

**Comment by dukedave** [comment]
Oh, I meant specifically the act of checking in a `.zen-deploy.yml` with placeholder values [over there](https://github.com/giftcardzen/main-app/pull/2795#discussion_r61498226).
No complaints here (aside from lack of docs, but, I can just ask :smile:).

**Pull request merged**